---
title: "R Notebook"
output: html_notebook
---

```{r}
#Load the library
library(vegan)
library(tidyverse)
library(devtools)
library(grid) 
library(gridExtra)
library(knitr)
library(png)
library(ggpubr) 
library(RColorBrewer)
library(ggpubr)
library(ggplot2)
library(dplyr)
library(phyloseq)
library(ggnewscale)
library(SRS)
library(MicrobiotaProcess)
library(FSA)
```

```{r}
#load the biom file and metadata
data <- import_biom(file.choose(),parseFunction=parse_taxonomy_default)
```


```{r}
#
metadata <- read.csv(file.choose(), header=TRUE, 
                     row.names=1 )
```


```{r}
names(data)
sample_names(data)
rownames(metadata)
```


```{r}
all(rownames(metadata) %in% sample_names(data))

sample_data <- sample_data(metadata)
```


```{r}
physeq2 <- merge_phyloseq(sample_data, data )
physeq2
View(physeq2@tax_table@.Data)
```


```{r}
physeq2@tax_table@.Data <- substring(data@tax_table@.Data, 4)
colnames(physeq2@tax_table@.Data)<- c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")
colnames(physeq2@tax_table@.Data)

#justbacteria <- subset_taxa(physeq2, !Kingdom %in% c("Eukaryota"))
#justbacteria2 <- subset_taxa(justbacteria, !Kingdom %in% c("Viruses"))
justbacteria3 <- subset_taxa(physeq2, !Phylum %in% c("Chordata"))
```


```{r}
#To know number of reads per sample and sort them 

#sample_sums(justbacteria3)
sort(sample_sums(justbacteria3))
sort(sample_sums(physeq2))

```


```{r}
#SRS
minControl=5595
keep=names(which(sample_sums(justbacteria3)>=minControl))
pruned_all=prune_samples(keep, justbacteria3)
any(taxa_sums(pruned_all) == 0)

pruned_all <- prune_taxa(taxa_sums(justbacteria3) > 0, justbacteria3)
any(taxa_sums(pruned_all) == 0)
```


```{r}
pruned_data_df = as.data.frame(otu_table(pruned_all))

SRS_OUTPUT <- SRS (pruned_data_df,minControl)

rownames(SRS_OUTPUT)=rownames(pruned_data_df)

table(sample_data(pruned_all)[ ,"Treatment"]) 
```


```{r}
# transform back into phyloseq object
#taxa = otu_table(SRS_OUTPUT, taxa_are_rows = TRUE)
otu_table(pruned_all)=taxa
any(taxa_sums(pruned_all) == 0)

pruned_all <- prune_taxa(taxa_sums(pruned_all) > 0, pruned_all)
any(taxa_sums(pruned_all) == 0)

SeqDepthPruned = sample_sums(pruned_all)
sample_data(pruned_all)$SeqDepthPruned = SeqDepthPruned

```


```{r}
#Taxonomic classification

genus_colors <-c("#A6CEE3", "#1F78B4", "#B2DF8A", "#33A02C",
                 "#FB9A99", "#E31A1C", "#FDBF6F", "#00FFFF",
                 "#FF7F00", "#CAB2D6","#8A7C64","#652926",
                 "#6A3D9A", "#B15928", "#FFC000")
 

all_genus <- tax_glom(pruned_all, "Genus", NArm = TRUE)
# Get top 15 genera
top15_genera <- names(sort(taxa_sums(all_genus), decreasing=TRUE))[1:15]
# Transform Taxa counts to relative abundance
all_genus_relabun <- transform_sample_counts(all_genus, function(OTU) OTU/sum(OTU) * 100)
# Extract the top 15 taxa
all_genus_top15 <- prune_taxa(top15_genera, all_genus_relabun)
# Convert into dataframe
taxa_abundance_table_genus <- psmelt(all_genus_top15)
#Plot top 15 genus
T= StackedBarPlot_genus_endo <- taxa_abundance_table_genus %>%
  ggplot(aes(x =Sample, y = Abundance, fill = Genus)) +
  geom_bar(stat = "identity") + theme_bw() +
  labs(x = "Treatment",
       y = "Relative Counts of reads") + 
  facet_grid(~ Treatment, scales='free_x',space="free_x")+
  theme(
    axis.text.x = element_text(size = 9, angle = 90, vjust = 0.5, hjust = 1),
    axis.text.y = element_text(size = 9),
    legend.text = element_text(size = 9),
    strip.text = element_text(size = 9)) +
  scale_fill_manual(values=genus_colors)
T

ggsave( "/Users/amanpreetkaur/Downloads/Bacterial_diversity.pdf", width = 8, height = 7, units="in", dpi=800)
```


```{r}
counts_lizard <- read.csv("/Users/amanpreetkaur/Library/CloudStorage/Box-Box/FunGen2023/Team2_ATAC-seq_HeatStress_Lizards/Results/kraken2/counts_lizard.csv")


counts_lizard$log <- log10(counts_lizard$counts)
before <- counts_lizard[counts_lizard$which %in% c( "total", "Lizard"), ]
After <- counts_lizard[counts_lizard$which %in% c("Human" , "prokaryotes", "Lizard"), ]

sample_level1 <- factor(before$which, level=c("total", "Lizard" ))
sample_level2 <- factor(After$which, level=c(  "Lizard", "prokaryotes", "Human"  ))

B = before %>%
  ggplot(aes(x =Sample , level = c(), y = counts, fill = sample_level1)) +
  geom_bar(stat = "identity", position = "dodge", width=0.8)  + #,position = "fill" ) + 
  scale_fill_manual(values= c( "Lizard"= "#B15928", 
                               "not-matched" = "#6A3D9A", 
                               "total"="#FFC000")) +
  labs(x = "Sample",
       y = "Log of read counts") + theme_bw() + facet_grid(~ Treatment, scales='free_x',space="free_x")+
  theme(
    axis.text.x = element_text(size = 7,  angle = 90,face="bold", vjust = 0.5, hjust = 1),
    axis.text.y = element_text(size = 7),
    legend.text = element_text(size = 7),
    strip.text = element_text(size = 7))
#element_text(face="bold", color="#993333",  size=14, angle=45)
B
ggsave( "/Users/amanpreetkaur/Downloads/total_reads.pdf", width = 3.5, height = 3.0, units="in", dpi=800)

A = After %>%
  ggplot(aes(x =Sample, y = log, fill = sample_level2)) +
  geom_bar(stat = "identity", position = "dodge", width=0.8)  + #,position = "fill" ) +
 scale_fill_manual(values= c("Lizard"= "#B15928",
                             "prokaryotes"= "#FFC000", 
                             "Human" = "#A6CEE3")) +
  labs(x = "Sample",
       y = "Log of read counts") + theme_bw() + facet_grid(~ Treatment, scales='free_x',space="free_x")+
  theme(
    axis.text.x = element_text(size = 7,  angle = 90,face="bold", vjust = 0.5, hjust = 1),
    axis.text.y = element_text(size = 7),
    legend.text = element_text(size = 7),
    strip.text = element_text(size = 7))
#element_text(face="bold", color="#993333",  size=14, angle=45)
A
#ggsave( "/Users/amanpreetkaur/Downloads/Bacterial_contamination.pdf", width = 3.8, height = 3.0, units="in", dpi=800)
```

```{r}
#normality test
shapiro.test(before$counts)
#significance test
kruskal.test(before$counts ~ before$Treatment)
dunnTest(before$counts ~ before$which)

